cg.go

```diff
@@ -0,0 +1,293 @@
+	floats.Scale(-1, direction)
+
+	cg.iterFromRestart++
+	var restart bool
+	// Check if the method should be restarted because too many iterations have
+	// been taken without a restart.
+	if cg.iterFromRestart%cg.restartAfter == 0 {
+		restart = true
+	}
+	// Check if the method should be restarted because the last two gradients
+	// point in almost opposite directions.
+	gNorm := floats.Norm(loc.Gradient, 2)
+	gDotGPrev := floats.Dot(loc.Gradient, cg.gradPrev)
+	if gDotGPrev <= -cg.AngleRestartThreshold*gNorm*cg.gradPrevNorm {
+		restart = true
+	}
+
+	// Compute the scaling factor β_k according to the given CG variant. Do
+	// this always, because cg.Variant may be keeping its inner state that
+	// needs to be updated at every iteration.
+	beta := cg.Variant.Beta(loc.Gradient, cg.gradPrev, cg.dirPrev)
+	if beta == 0 {
+		// β_k == 0 means that the steepest descent direction will be taken, so
+		// indicate that the method is in fact being restarted.
+		restart = true
+	}
+	if !restart {
+		// The method is not being restarted, so update the descent direction.
+		floats.AddScaled(direction, beta, cg.dirPrev)
+	} else {
+		// Otherwise reset to 0 the counter of iterations taken since the last
+		// restart.
+		cg.iterFromRestart = 0
+	}
+
+	// Get the initial line search step size from the StepSizer, even if the
+	// method was restarted, because StepSizers need to see every iteration.
+	stepSize = cg.InitialStep.StepSize(loc, direction)
+	if restart {
+		// The method was restarted and since the steepest descent direction is
+		// not related to the previous direction, discard the estimated step
+		// size from cg.InitialStep and use step size of 1 instead.
+		stepSize = 1
+	}
+
+	copy(cg.gradPrev, loc.Gradient)
+	copy(cg.dirPrev, direction)
+	cg.gradPrevNorm = gNorm
+	return stepSize
+}
+
+// FletcherReeves implements the Fletcher-Reeves variant of the CG method that
+// computes the scaling parameter β_k according to the formula
+// β_k = |∇f_{k+1}|^2 / |∇f_k|^2.
+type FletcherReeves struct {
+	gPrevNorm float64
+}
+
+func (fr *FletcherReeves) Init(loc Location) {
+	fr.gPrevNorm = floats.Norm(loc.Gradient, 2)
+}
+
+func (fr *FletcherReeves) Beta(grad, gradPrev, dirPrev []float64) (beta float64) {
+	gNorm := floats.Norm(grad, 2)
+	beta = gNorm * gNorm / fr.gPrevNorm / fr.gPrevNorm
+
+	fr.gPrevNorm = gNorm
+	return beta
+}
+
+// PolakRibierePolyak implements the Polak-Ribiere-Polyak variant of the CG
+// method that computes the scaling parameter β_k according to the formula
+// β_k = max(0, ∇f_{k+1}·y_k / |∇f_k|^2),
+// where y_k = ∇f_{k+1} - ∇f_k.
+type PolakRibierePolyak struct {
+	gPrevNorm float64
+}
+
+func (pr *PolakRibierePolyak) Init(loc Location) {
+	pr.gPrevNorm = floats.Norm(loc.Gradient, 2)
+}
+
+func (pr *PolakRibierePolyak) Beta(grad, gradPrev, dirPrev []float64) (beta float64) {
+	gNorm := floats.Norm(grad, 2)
+	gDotGPrev := floats.Dot(grad, gradPrev)
+	beta = (gNorm*gNorm - gDotGPrev) / pr.gPrevNorm / pr.gPrevNorm
```

Comment:
Actually, I take it back. probably best as
beta = (norm * norm - dot) / (pr.prevNorm * pr.prevNorm)